Feature/summit jsrun tests

vivek-bala opened the pull request:
Opening a PR so everyone can take a look at the test suite. Let me know what you think.

There is one aspect of the test that I am still determining how to do: tests where we run MPI application where each process needs to use a distinct node, we need to assert that in fact two different nodes were used (tests 5 and 6). If you have any ideas, feel free to share.

itomaldonado commented:
Can one of the admins verify this patch?

mturilli commented on tests/test_launch_methods/test_jsrun/jobs.sh:
```diff
@@ -0,0 +1,83 @@
+
+    cp $actual results_$d/$tid/jsrun_output_dup.log
+    # Remove Node ID from actual output
+    sed -i 's/Node ....../Node /g' results_$d/$tid/jsrun_output_dup.log
```
Why do you remove the IDs?

andre-merzky replied:
Presumably because the differ from run to run, and are thus difficult to verify against a static output?

vivek-bala replied:
Yes, that's correct. The node IDs differ between runs. The same can hold for HW threads in some tests.

mturilli commented on tests/test_launch_methods/test_jsrun/jobs.sh:
```diff
@@ -0,0 +1,83 @@
+    rm results_$d/$tid -rf
+    mkdir results_$d/$tid
+
+    # Record resource file, omp env, and cmd in log
```
What omp stands for?

iparask replied:
It is for OPENMP

vivek-bala replied:
Yes it stands for OpenMP. The value is being extracted from the scripts and executed in the current session.

mturilli commented on tests/test_launch_methods/test_jsrun/test_1/cmds:
```diff
@@ -0,0 +1,3 @@
+# Task needs to be assigned node 1, physical core 1, HW thread 1
+export OMP_NUM_THREADS=1
+jsrun -a 1 -U resource-file /ccs/proj/csc190/Hello_jsrun/jsrun_layout
```
Add newline at te end of file?

vivek-bala replied:
Done. Thanks.

iparask commented on tests/test_launch_methods/test_jsrun/jobs.sh:
```diff
@@ -0,0 +1,83 @@
+d=`date +%D | sed 's/\//_/g'`
+host=`/bin/hostname`
+#dest="rct-tests@googlegroups.com"
+dest="vivek.bala@rutgers.edu"
```
I agree with Matteo

vivek-bala replied:
Done. Makes sense. Thanks.

iparask commented on tests/test_launch_methods/test_jsrun/jobs.sh:
```diff
@@ -0,0 +1,83 @@
+    # Set omp env, create resource file and run cmd
+    $omp_env
+    echo $res_set >> $path/resource-file
+    cmd=`echo $cmd | sed "s#resource-file#$path/resource-file#g"`
```
Since there is a resource file in each folder why you write it to a different file and then remove it?

vivek-bala replied:
The content of the resource file is appended to the output log inside each folder. The resource file itself
is created in ``path`` and then deleted at the end of the individual test.

iparask replied:
Yes, but why are you doing all that, when you can just  use the `res_set` file as input?

vivek-bala replied:
Fair point. Changed and tested.

iparask commented on tests/test_launch_methods/test_jsrun/test_5/cmds:
```diff
@@ -0,0 +1,3 @@
+# Task needs to be assigned nodes 1 and 2, physical core 1 of each node, HW thread 1 of each physical core
+export OMP_NUM_THREADS=1
+jsrun -a 1 -U resource-file /ccs/proj/csc190/Hello_jsrun/jsrun_layout
```
You are missing `verbose`, otherwise the node name will never appear in the output

vivek-bala replied:
I am not sure I follow. These were the exact commands that were run and I do get the node name/ID. Let me know if you think I missed something.

iparask commented:
Why don't you take the node list (`$LSB_HOST`) and keep the unique values apart from the first?

andre-merzky commented:
@vivek-bala : action needed wrt. comments.

mturilli commented on tests/test_launch_methods/test_jsrun/test_1/res_set:
```diff
@@ -0,0 +1 @@
+RS 0: { host: 1, cpu: 0}
```
Should we use ERF?

vivek-bala replied:
Yes I agree. @iparask let me know how we can proceed on that.

iparask replied:
@vivek-bala I'll make a commit now that you are done

iparask commented on tests/test_launch_methods/test_jsrun/jobs.sh:
```diff
@@ -0,0 +1,79 @@
+    if ! diff $expected results_$d/$tid/jsrun_output_dup.log; then
+        op="$tid failed: Output location = results_$d/$tid/jsrun_output.log, Diff location = results_$d/$tid/full_output.log"
+        echo "Diff between expected and actual (bar node ID):" >> results_$d/$tid/full_output.log
+        diff $expected results_$d/$tid/jsrun_output_dup.log >> results_$d/$tid/full_output.log
```
Can you put the absolute path here? I am not sure from the email where the log file is. As a result, I cannot check it

mturilli replied:
@vivek-bala any follow up on this?

iparask commented on tests/test_launch_methods/test_jsrun/test_1/cmds:
```diff
@@ -0,0 +1,3 @@
+# Task needs to be assigned node 1, physical core 1, HW thread 1
+export OMP_NUM_THREADS=1
+jsrun -a 1 -U resource-file /ccs/proj/csc190/Hello_jsrun/jsrun_layout
```
I suggest the following:
```suggestion
jsrun -a 1 -U resource-file /ccs/proj/csc190/Hello_jsrun/jsrun_layout | sort
```

This way all outputs will be sorted and easier to compare with the file you produce. Especially with MPI execution, where the print order is not guaranteed.

iparask commented:
May also I suggest that it moves under the `test/lm/` folder as integration tests?

vivek-bala commented:
Moved the tests and made some changes to the test assertions that are more extensible

iparask commented:
This is an integration test, right? May I suggest then that there is a folder named `integration` under `lm` and `test_jsurn` to be named just `jsrun`?
It will fall very nicely with the `unit` folder that denotes the unit tests!

vivek-bala commented:
Thanks Giannis. Good point. Moved it into another folder integration tests. I renamed some folders with 'test_' so that they can be auto-discovered by pytest.

andre-merzky commented:
Merging now.  I think that Ioannis' request for absolute paths is still open, but I don't want to delay further.

Thanks all!
